src/python/WMCore/MicroService/MSPileup/DataStructs/MSPileupObj.py

```diff
@@ -0,0 +1,174 @@
+        self.deactivatedOn = pdict.get('deactivatedOn', gmtimeSeconds())
+        self.active = pdict.get('active', False)
+        self.pileupSize = pdict.get('pileupSize', 0)
+        self.ruleList = pdict.get('ruleList', [])
+
+        self.data = {
+            'pileupName': self.pileupName,
+            'pileupType': self.pileupType,
+            'insertTime': self.insertTime,
+            'lastUpdateTime': self.lastUpdateTime,
+            'expectedRSEs': self.expectedRSEs,
+            'currentRSEs': self.currentRSEs,
+            'fullReplicas': self.fullReplicas,
+            'campaigns': self.campaigns,
+            'containerFraction': self.containerFraction,
+            'replicationGrouping': self.replicationGrouping,
+            'activatedOn': self.activatedOn,
+            'deactivatedOn': self.deactivatedOn,
+            'active': self.active,
+            'pileupSize': self.pileupSize,
+            'ruleList': self.ruleList}
+        if not self.validate(self.data):
+            msg = 'MSPileup input is invalid'
+            raise Exception(msg)
+
+    def __str__(self):
+        """
+        Return human readable representation of pileup data
+        """
+        return json.dumps(self.data, indent=4)
+
+    def getPileupData(self):
+        """
+        Get pileup data
+        """
+        return self.data
+
+    def validate(self, pdict):
+        """
+        Data validator according to its schema
+
+        :param pdict: input data dictionary
+        :return: boolean, result of validation
+        """
+        docSchema = self.schema()
+        for key, val in pdict.items():
+            if key not in docSchema:
+                msg = f"Failed to validate {key}, not found in {docSchema}"
+                self.logger.error(msg)
+                return False
+            _, stype = docSchema[key]  # expected data type for our key
+            if not isinstance(val, stype):
+                dtype = str(type(val))     # obtained data type of our value
+                msg = f"Failed to validate: {key}, expect data-type {stype} got type {dtype}"
+                self.logger.error(msg)
+                return False
+            if key == 'pileupName' and not dataset(val):
+                msg = f"pileupName value {val} does not match dataset pattern"
+                self.logger.error(msg)
+                return False
+            if key == "pileupType" and val not in ['classic', 'prefix']:
+                msg = f"pileupType value {val} is neither of ['classic', 'prefix']"
+                self.logger.error(msg)
+                return False
+            if key == 'replicationGrouping' and val not in ['DATASET', 'ALL']:
+                msg = f"replicationGrouping value {val} is neither of ['DATASET', 'ALL']"
+                self.logger.error(msg)
+                return False
+            if key == 'containerFraction' and isinstance(val, float) and (val > 1 or val < 0):
```

Comment:
Isn't data type already validated in the beginning of this for loop?

Reply:
it is, so I removed `isinstance` call then.